lib/src/stream_extensions.dart

```diff
@@ -78,4 +80,71 @@ extension StreamExtensions<T> on Stream<T> {
       ..onCancel = subscription.cancel;
     return controller.stream;
   }
+
+  /// Invoke [each] for each item in this stream, and wait for the [Future]
+  /// returned by [each] to be resolved. Running no more than [concurrency]
+  /// number of [each] calls at the same time.
+  ///
+  /// This function will wait for the futures returned by [each] to be resolved
+  /// before completing. If any [each] invocation throws, [boundedForEach] stop
+  /// subsequent calls
+  ///
+  /// will
+  /// call [onError], if [onError] throws (it does)
+  /// continue subsequent [each] calls, ignore additional errors and throw the
+  /// first error encountered.
+  Future<void> boundedForEach(
+    int concurrency,
+    FutureOr<void> Function(T item) each, {
+    FutureOr<void> Function(Object e, StackTrace? st) onError = _throwOnError,
+  }) async {
+    Object? firstError;
+    StackTrace? firstStackTrace;
+
+    var running = 0;
+    final wakeUp = ConditionVariable();
+
+    try {
+      var doBreak = false;
+      await for (final item in this) {
+        running += 1;
+        scheduleMicrotask(() async {
```

Comment:
Don't think the `scheduleMicrotask` is needed. 
The code is running inside an `await for`, so we're definitely past the initial synchronous part of the `async` functions, and already running inside an event callback.

You can still run the code inside a separate async function, as `() async { ... }();`, without using `scheduleMicrotask`. (Might need to use `unawaited`, but passing an async function to `scheduleMicrotask` is already a code-smell, since it doesn't look at the future anyway.)